p-ata/src/processor.rs

```diff
@@ -0,0 +1,595 @@
+            data: &INITIALIZE_IMMUTABLE_OWNER_DATA,
+        };
+        cpi::invoke(&init_immutable_owner_ix, &[associated_token_account])?;
+    }
+
+    // Initialize account via InitializeAccount3.
+    let initialize_account_instr_data = build_initialize_account3_data(wallet.key());
+    let initialize_account_metas = &[
+        AccountMeta {
+            pubkey: associated_token_account.key(),
+            is_writable: true,
+            is_signer: false,
+        },
+        AccountMeta {
+            pubkey: mint_account.key(),
+            is_writable: false,
+            is_signer: false,
+        },
+    ];
+    let init_ix = Instruction {
+        program_id: token_program.key(),
+        accounts: initialize_account_metas,
+        data: &initialize_account_instr_data,
+    };
+
+    cpi::invoke(&init_ix, &[associated_token_account, mint_account])?;
+    Ok(())
+}
+
+/// Check if a given address is off-curve (not a valid Ed25519 curve point).
+///
+/// Returns `true` if the address is off-curve, `false` if on-curve.
+///
+/// - **On-chain (Solana)**: Uses `sol_curve_validate_point` syscall
+/// - **Tests**: Uses curve25519-dalek to replicate on-chain behavior  
+/// - **Other builds**: Returns `false`
+#[inline(always)]
+pub(crate) fn is_off_curve(_address: &Pubkey) -> bool {
+    #[cfg(target_os = "solana")]
+    {
+        const ED25519_CURVE_ID: u64 = 0;
+
+        let point_addr = _address.as_ref().as_ptr();
+
+        // SAFETY: We're passing valid pointers to the syscall
+        // The syscall directly returns the validation result:
+        // - 0 means point is ON the curve (valid)
+        // - 1 means point is OFF the curve (invalid)
+        // - any other value means error
+        let syscall_result = unsafe {
+            sol_curve_validate_point(ED25519_CURVE_ID, point_addr, core::ptr::null_mut())
+        };
+
+        syscall_result == 1
+    }
+    #[cfg(all(not(target_os = "solana"), test))]
+    {
+        // Host build (tests, benches): replicate the on-chain `sol_curve_validate_point` logic
+        // using curve25519-dalek. A pubkey is "off-curve" if it cannot be decompressed into
+        // an Edwards point **or** it decomposes to a small-order point
+        // (matches Solana’s runtime rules).
+
+        use curve25519_dalek::edwards::CompressedEdwardsY;
+
+        let mut bytes = MaybeUninit::<[u8; 32]>::uninit();
+        let bytes_ptr = bytes.as_mut_ptr() as *mut u8;
+        // SAFETY: We initialize all 32 bytes before calling assume_init()
+        let bytes = unsafe {
+            core::ptr::copy_nonoverlapping(_address.as_ref().as_ptr(), bytes_ptr, 32);
+            bytes.assume_init()
+        };
+        let compressed = CompressedEdwardsY(bytes);
+
+        match compressed.decompress() {
+            None => true,                    // invalid encoding → off-curve
+            Some(pt) => pt.is_small_order(), // small-order = off-curve, otherwise on-curve
```

Comment:
This might be my ignorance, but where is the `is_small_order` check coming from? We typically just check to see if it decompresses. Here's the syscall code: https://github.com/anza-xyz/agave/blob/d30889f6ddd58d68c3a5448fcd1aebaa96789a99/curves/curve25519/src/edwards.rs#L82